python/ray/util/client/server/dataservicer.py

```diff
@@ -33,23 +48,47 @@ def Datapath(self, request_iterator, context):
             logger.error("Client connecting with no client_id")
             return
         logger.debug(f"New data connection from client {client_id}: ")
+        if not self._init(client_id, context):
+            return
         try:
-            for req in request_iterator:
+            request_queue = queue.Queue()
+            queue_filler_thread = Thread(
+                target=fill_queue,
+                daemon=True,
+                args=(request_iterator, request_queue))
+            queue_filler_thread.start()
+            """For non `async get` requests, this loop yields immediately
+            For `async get` requests, this loop:
+                 1) does not yield, it just continues
+                 2) When the result is ready, it yields
+            """
+            for req in iter(request_queue.get, None):
+                if isinstance(req, ray_client_pb2.DataResponse):
+                    # Early shortcut if this is the result of an async get.
+                    yield req
+                    continue
                 resp = None
                 req_type = req.WhichOneof("type")
                 if req_type == "init":
-                    resp = self._init(req.init, client_id)
-                    if resp is None:
-                        context.set_code(grpc.StatusCode.RESOURCE_EXHAUSTED)
-                        return
+                    resp_init = self.basic_service.Init(req.init)
+                    resp = ray_client_pb2.DataResponse(init=resp_init, )
                     logger.debug(f"Accepted data connection from {client_id}. "
                                  f"Total clients: {self.num_clients}")
                     accepted_connection = True
                 else:
                     assert accepted_connection
```

Comment:
is this assert necessary?

Reply:
Yes, because it ensures init was called first.

Reply:
but why do we need it? do we want to fail if accepted_connection is False? Shouldn't we instead check if it is true and if not raise an exception?

Reply:
This has been removed as of my commit from yesterday.